Bug not recovering after an error

EranOfer opened the pull request:
In the producer made each produce request get its own response with a new offset the corrosponds to the request.
1. In the producer made Ack=0 requests return a single response with offset -1
2. In TcpSocket made a finite number of retries. After all retries fail, putting exceptions in each response for each request.
3. In TcpSocket. Removed dependancy between current tasks and future tasks

Commit:
fix test ShouldRetryWhenReceiveAnRetryErrorCode
add maxRetryAttempt to KafkaMetadataProvider

Commit:
move the partitionId to central place in ManualConsumerTests

Commit:
refractor and add test

Commit:
Start Refactoring
change Behavior  of ack Level  0
still work in progress not done refactoring

Commit:
1. In the producer made each produce request get its own response with a new offset the corrosponds to the request.
2. In the producer made Ack=0 requests return a single response with offset -1
3. In TcpSocket made a finite number of retries. After all retries fail, putting exceptions in each response for each request.
4. In TcpSocket. Removed dependancy between current tasks and future tasks

Commit:
Add test WhenNoConnectionThrowSocketExceptionAfterMaxRetry

Commit:
add ProduserAckLevel test

Commit:
refactor comments

Commit:
refactor comments

Commit:
add test
ProducerLastResposeOffsetAckLevel1ShouldBeEqualsToLastOffset
ProducerAckLevel1ResponseOffsetShouldBeEqualToLastOffset
ConsumeByOffsetShouldGetSameMessageProducedAtSameOffset

Commit:
trying to add async consumer .
not done!

Commit:
Revert "trying to add async consumer . not done!"

This reverts commit ebfaeb8c8d97e5eeaafbf0fd468f6044c7779422.

yoniabrahamy commented on src/KafkaNetClient/Consumer.cs:
```diff
@@ -124,15 +126,15 @@ private Task ConsumeTopicPartitionAsync(string topic, int partitionId)
                         try
                         {
+                            //after error
+                            if (refreshMetaData)
```
change name to "needToRefreshMetadata"

yoniabrahamy commented on src/KafkaNetClient/Consumer.cs:
```diff
@@ -180,13 +184,14 @@ private Task ConsumeTopicPartitionAsync(string topic, int partitionId)
                                     }
 
-                                    var nextOffset = response.Messages.Max(x => x.Meta.Offset) + 1;
+                                    var nextOffset = response.Messages[response.Messages.Count - 1].Meta.Offset + 1;
```
Why not response.Messages.Last()?

yoniabrahamy commented on src/KafkaNetClient/KafkaTcpSocket.cs:
```diff
@@ -138,92 +142,112 @@ private Task<byte[]> EnqueueReadTask(int readSize, CancellationToken cancellatio
+            var sendTask = ProcessNetworkstreamTasksReadTask(netStream);
+            await Task.WhenAny(readTask, sendTask);
+            if (_disposeToken.IsCancellationRequested) return;
+            await ThrowTaskException(readTask);
```
Maybe change the function name to something that implies that an exception will not be necessarily thrown. Maybe ThrowTaskExceptionIfFaulted

yoniabrahamy commented on src/KafkaNetClient/KafkaTcpSocket.cs:
```diff
@@ -389,6 +421,10 @@ private static Exception FindExceptionInLastReadOrWriteTasks(Task lastWriteTask,
                 }
 
                 await Task.Delay(TimeSpan.FromMilliseconds(reconnectionDelay), _disposeToken.Token).ConfigureAwait(false);
+                if (_maxRetry < attempts)
```
Move this condition inside the catch so we don't accidentally throw an exception if we succeeded.

yoniabrahamy commented on src/KafkaNetClient/Producer.cs:
```diff
@@ -274,31 +278,57 @@ private bool IsNotDisposedOrHasMessagesToProcess()
-                catch
+                catch (Exception)
+                {
+                    //TODO:Add more log
```
Well then, Where is the log?

yoniabrahamy commented on src/kafka-tests/Fakes/BrokerRouterProxy.cs:
```diff
@@ -70,7 +70,7 @@ public IBrokerRouter Create()
         }
 
-        public static async Task<MetadataResponse> DefaultMetadataResponse()
+        public static async Task<MetadataResponse> CreateMetaResponseWith2Broker()
```
The name is 2Unclear4me. Why only two brokers? what's going on?

yoniabrahamy commented on src/kafka-tests/Fakes/BrokerRouterProxy.cs:
```diff
@@ -119,5 +119,54 @@ public IBrokerRouter Create()
                 };
         }
+
+        public static async Task<MetadataResponse> CreateMetaResponseWith1Broker()
```
change to CreateMetadataResponseWithSingleBroker

yoniabrahamy commented on src/kafka-tests/Unit/KafkaConnectionTests.cs:
```diff
@@ -37,7 +38,7 @@ public void Setup()
         public async Task ShouldStartReadPollingOnConstruction()
         {
-            using (var socket = new KafkaTcpSocket(_log, _kafkaEndpoint))
+            using (var socket = new KafkaTcpSocket(_log, _kafkaEndpoint, 5))
```
Why not use _maxRetry?

yoniabrahamy commented on src/kafka-tests/Unit/KafkaConnectionTests.cs:
```diff
@@ -49,7 +50,7 @@ public void Setup()
         {
             var expectedUrl = _kafkaEndpoint;
-            using (var socket = new KafkaTcpSocket(_log, expectedUrl))
+            using (var socket = new KafkaTcpSocket(_log, expectedUrl, 5))
```
Same here..

yoniabrahamy commented on src/kafka-tests/Unit/KafkaTcpSocketTests.cs:
```diff
@@ -36,7 +40,7 @@ public KafkaTcpSocketTests()
         public void KafkaTcpSocketShouldConstruct()
         {
-            using (var test = new KafkaTcpSocket(_log, _fakeServerUrl))
+            using (var test = new KafkaTcpSocket(_log, _fakeServerUrl, 5))
```
Also here.

Commit:
code review comment

Commit:
Merge branch 'BugNotRecoveringAfterAnError' of https://github.com/EranOfer/KafkaNetClient.git

Commit:
add  log to producer

Commit:
code review comment

BDeus commented:
hi,
i try to debug the code but i have some regression with this branch when i intentionnaly shutdown a kafka broker.
I found an infinite loop with the asynccollection and the StartReadStreamPoller.
StartReadStreamPoller add an item in the readqueue and the asynccollection remains in  DrainAndApply function (in the while dequeue loop)

EranOfer commented:
hi,
Thanks alot for your help.
I will check this matter tomorrow.
If you could add a failed test it will help.

BDeus commented:
i wrote a failed test and the patch of the asynccollection
do you want me to push it in your branch? or just copy and paste here?

EranOfer commented:
I will glad if you do both.
Copy and paste here and create pull request on my local branch 
https://github.com/EranOfer/KafkaNetClient/tree/BugNotRecoveringAfterAnError .
I will go over the code and merge it .

EranOfer commented:
and again thanks alot for your help :+1:

randa1 commented:
Actually, @EranOfer, I think it's better if @BDeus will create a pull request to this repository (in master branch) and not to your local one. Later you can pull the changes to your fork. It's better to handle all contributions in here than to create a multi-repository hierarchy.

Commit:
Create:Statistics Tracker Options that it can be disable
Fix: memory leak on Statistics Tracker when sending large message
Fix: critical exception in Statistics Tracker (Concurrent Circular Buffer OutOfRangeException)

Commit:
remove error message

Commit:
fix bug -infinite loop with the asynccollection and the StartReadStreamPoller.

this code need more refactor.

BDeus commented:
Finally i have create a pull request on @EranOfer branch because i had some refactoring in my branch before the patch.
I prefer than @EranOfer merge it before in is own branch.

Commit:
Add Test for AsyncCollection.DrainAndApply and patch a possible bug

Commit:
Revert "fix bug -infinite loop with the asynccollection and the StartReadStreamPoller."

This reverts commit 7bb82153a45aa5f9272a984e1668e246e2b87d27.

Commit:
Merge pull request #5 from BDeus/BugNotRecoveringAfterAnError

Add Test for AsyncCollection.DrainAndApply and patch a possible bug

Commit:
Merge branch 'BugNotRecoveringAfterAnError' of https://github.com/EranOfer/KafkaNetClient.git
